detangle_client_base_classes.rst

```diff
@@ -0,0 +1,224 @@
+
+
+
+Analysis
+--------
+
+Client credentials
+~~~~~~~~~~~~~~~~~~
+
+- All clients need to be able to take credentials as a ctor argument.  If
+  not passed, they need to be able *at some point before credentials are
+  needed* to infer them from the environment.
+
+- Some clients may not need to have explicit credentials.  E.g., the ``storage``
+  client has a custom classmethod factory, ``create_anonymous_client``,
+  which uses ``google.auth.credentials.AnonymousCredentials``.
+
+- Eager construction / lookup of credentials in the ctor is probably the
+  wrong thing, but does have the benefit of ensuring thread-saftey for
+  credentials after construction of the client.
+
+- In addition to using credentials to initialize their HTTP or GRPC
+  transport wrapprs, some clients may need to use credentials for other
+  purposes.  E.g., the ``storage`` client's credentials are used to create
+  signed URLs providing temporary anonymous access to upload files to a
+  bucket.
+
+Client project ID
+~~~~~~~~~~~~~~~~~
+
+- All clients need to be able to take a project ID as a ctor argument.  If
+  not passed, they need to be able *at some point before a project ID is
+  needed* to infer it from the environment (or the credentials).
+ 
+- Some clients may not need an explicit project ID.  E.g., that
+  the ``storage`` client's ``create_anonymous_client`` passes a sentinel
+  value for the project ID.  It's constructor also does a convoluted dance
+  for the case that ``project`` is explicitly passed as ``None``, restoring
+  that value after calling the base class' ``__init__``.
+
+- Eager construction / lookup of the project ID in the ctor is probably the
+  wrong thing, but does have the benefit of ensuring thread-saftey for
+  the value after construction of the client.
+
+- The patterns for inferring the project ID are convoluted, and maybe
+  buggy (e.g., the known flakiness of the GCE metadata server).
+  They also interact poorly with generalized support for API emulators (e.g.,
+  the contortions in the ``datastore`` client to support the emulator's
+  environment variable).
+
+HTTP-based Clients
+~~~~~~~~~~~~~~~~~~~
+
+- Currently, HTTP-based clients hold / create an object which emulates the
+  ``request.Session`` API, and make that object available to their connection
+  instances (API wrappers) via their ``_http`` attribute.
+
+- I believe it would be cleaner to have the instance passed to the connection
+  objects' constructors, rather than having them use the attribute.
+
+- The ``storage`` client contorts around the ``_http`` attribute in its
+  ``test_batch`` unit tests:  get rid of that!
+
+- In addition to its "normal" connection wrapper, the ``bigquery`` client
+  passes its ``_http`` attribute to the "upload" objects used for loading
+  table data to GCS in its ``load_table_from_file`` method.
+
+- Clients which are passed on already-created ``requests.Session``-workalike
+  should *not* need / require / infer credentials.
```

Comment:
Is that also true in the case cited above, using _storage_ as an example, where the credentials are used for other purposes than initializing the transport wrappers?

Reply:
270b0fc